http/cookie_map.ts

```diff
@@ -0,0 +1,841 @@
+ * console.log(await cookies.get("foo")); // logs "bar"
+ * // the cookie will be automatically signed using the supplied key ring.
+ * await cookies.set("session", "1234567");
+ *
+ * const response = new Response("hello", {
+ *   headers: mergeHeaders({
+ *     "content-type": "text/plain",
+ *   }, cookies),
+ * });
+ * ```
+ *
+ * In addition, if you have a {@linkcode Response} or {@linkcode Headers} for a
+ * response at construction of the cookies object, they can be passed and any
+ * set cookies will be added directly to those headers:
+ *
+ * ```ts
+ * import { CookieMap } from "https://deno.land/std@$STD_VERSION/http/cookie_map.ts";
+ *
+ * const request = new Request("https://localhost/", {
+ *   headers: { "cookie": "foo=bar; bar=baz;"}
+ * });
+ *
+ * const response = new Response("hello", {
+ *   headers: { "content-type": "text/plain" },
+ * });
+ *
+ * const cookies = new CookieMap(request, { response });
+ * console.log(cookies.get("foo")); // logs "bar"
+ * cookies.set("session", "1234567");
+ * ```
+ *
+ * @module
+ */
+
+export interface CookieMapOptions {
+  /** The {@linkcode Response} or the headers that will be used with the
+   * response. When provided, `Set-Cookie` headers will be set in the headers
+   * when cookies are set or deleted in the map.
+   *
+   * An alternative way to extract the headers is to pass the cookie map to the
+   * {@linkcode mergeHeaders} function to merge various sources of the
+   * headers to be provided when creating or updating a response.
+   */
+  response?: Headered | Headers;
+  /** A flag that indicates if the request and response are being handled over
+   * a secure (e.g. HTTPS/TLS) connection. Defaults to `false`. */
+  secure?: boolean;
+}
+
+export interface CookieMapSetDeleteOptions {
+  /** The domain to scope the cookie for. */
+  domain?: string;
+  /** When the cookie expires. */
+  expires?: Date;
+  /** A flag that indicates if the cookie is valid over HTTP only. */
+  httpOnly?: boolean;
+  /** Do not error when signing and validating cookies over an insecure
+   * connection. */
+  ignoreInsecure?: boolean;
+  /** Overwrite an existing value. */
+  overwrite?: boolean;
+  /** The path the cookie is valid for. */
+  path?: string;
+  /** Override the flag that was set when the instance was created. */
+  secure?: boolean;
+  /** Set the same-site indicator for a cookie. */
+  sameSite?: "strict" | "lax" | "none" | boolean;
+}
+
+/** An object which contains a `headers` property which has a value of an
+ * instance of {@linkcode Headers}, like {@linkcode Request} and
+ * {@linkcode Response}. */
+export interface Headered {
+  headers: Headers;
+}
+
+export interface Mergable {
+  [cookieMapHeadersInitSymbol](): [string, string][];
+}
+
+export interface SecureCookieMapOptions {
+  /** Keys which will be used to validate and sign cookies. The key ring should
+   * implement the {@linkcode KeyRing} interface. */
+  keys?: KeyRing;
```

Comment:
Does it make sense to make this optional when we have non-secure alternative `CookieMap`? @kitsonk

Reply:
I personally think so, because the APIs are sync/async. So if you want to write your code to always use the async APIs but still only sometimes pass keys, you would just always use the secure version.

This is what I would do with oak, as swapping between the two implementations is a problem.

Reply:
Ok. Makes sense to me